File: backend/utils/duplicate_payment_notifications.py
"""
Duplicate Payment Notification System
Handles notifications when duplicate payments are detected across all payment streams
"""

import logging
from datetime import datetime
import pytz
from config import get_db_conn
from email_utils import send_simple_email
from fcm_service_fallback import fcm_service_fallback
from utils.security import decrypt_sensitive_data

logger = logging.getLogger(__name__)

def send_duplicate_payment_notifications(bl_id, bl_number, customer_username, customer_email, 
                                       payment_amount, payment_source, original_payment_date=None):
    """
    Send comprehensive duplicate payment notifications across all channels
    
    Args:
        bl_id: Bill of lading ID
        bl_number: Bill of lading number
        customer_username: Customer username
        customer_email: Customer email (encrypted)
        payment_amount: Amount of duplicate payment
        payment_source: Source of payment (webhook, email, bank_import, whatsapp)
        original_payment_date: Date of original payment (optional)
    """
    try:
        # Decrypt customer email
        decrypted_email = decrypt_sensitive_data(customer_email) if customer_email else None
        
        # Get customer phone number for WhatsApp
        customer_phone = get_customer_phone(customer_username)
        
        # Get customer FCM tokens
        fcm_tokens = get_customer_fcm_tokens(customer_username)
        
        # Send notifications
        send_fcm_duplicate_notification(fcm_tokens, bl_number, payment_amount, payment_source)
        send_email_duplicate_notification(decrypted_email, bl_number, payment_amount, payment_source, original_payment_date)
        send_whatsapp_duplicate_notification(customer_phone, bl_number, payment_amount, payment_source)
        send_staff_refund_alert(bl_number, customer_username, payment_amount, payment_source)
        
        logger.info(f"Duplicate payment notifications sent for BL {bl_number} (ID: {bl_id})")
        
    except Exception as e:
        logger.error(f"Error sending duplicate payment notifications for BL {bl_number}: {e}")

def get_customer_phone(username):
    """Get customer phone number from database"""
    try:
        conn = get_db_conn()
        cur = conn.cursor()
        cur.execute("SELECT customer_phone FROM users WHERE username = %s", (username,))
        result = cur.fetchone()
        cur.close()
        conn.close()
        
        if result and result[0]:
            return decrypt_sensitive_data(result[0])
        return None
    except Exception as e:
        logger.error(f"Error getting customer phone for {username}: {e}")
        return None

def get_customer_fcm_tokens(username):
    """Get customer FCM tokens from database"""
    try:
        conn = get_db_conn()
        cur = conn.cursor()
        cur.execute("SELECT token FROM fcm_tokens WHERE username = %s AND is_active = true", (username,))
        results = cur.fetchall()
        cur.close()
        conn.close()
        
        return [row[0] for row in results] if results else []
    except Exception as e:
        logger.error(f"Error getting FCM tokens for {username}: {e}")
        return []

def send_fcm_duplicate_notification(fcm_tokens, bl_number, payment_amount, payment_source):
    """Send FCM push notification to user about duplicate payment"""
    if not fcm_tokens:
        logger.info("No FCM tokens available for duplicate payment notification")
        return
        
    try:
        title = "⚠️ Duplicate Payment Detected"
        body = f"Your payment of ${payment_amount:.2f} for BL {bl_number} has already been processed. No action needed."
        
        data = {
            'type': 'duplicate_payment',
            'bl_number': bl_number,
            'payment_amount': str(payment_amount),
            'payment_source': payment_source,
            'timestamp': datetime.now().isoformat()
        }
        
        fcm_service_fallback.send_notification(
            tokens=fcm_tokens,
            title=title,
            body=body,
            data=data
        )
        
        logger.info(f"FCM duplicate payment notification sent to {len(fcm_tokens)} tokens")
        
    except Exception as e:
        logger.error(f"Error sending FCM duplicate payment notification: {e}")

def send_email_duplicate_notification(email, bl_number, payment_amount, payment_source, original_payment_date=None):
    """Send email notification to user about duplicate payment"""
    if not email:
        logger.info("No email available for duplicate payment notification")
        return
        
    try:
        subject = "⚠️ Duplicate Payment Alert - No Action Required"
        
        # Format original payment date
        original_date_str = ""
        if original_payment_date:
            if isinstance(original_payment_date, str):
                original_date_str = f"Original payment was processed on: {original_payment_date}"
            else:
                original_date_str = f"Original payment was processed on: {original_payment_date.strftime('%Y-%m-%d %H:%M:%S')}"
        
        body = f"""
Dear Customer,

We have detected a duplicate payment attempt for your shipment.

**Payment Details:**
- Bill of Lading: {bl_number}
- Duplicate Amount: ${payment_amount:.2f}
- Payment Source: {payment_source.title()}
{original_date_str}

**Important:** Your original payment has already been processed successfully. This duplicate payment will not be charged to your account.

**No action is required from you.** Your shipment processing continues as normal.

If you have any questions or concerns, please contact our support team.

Best regards,
Terry Ray Logistics Team
        """
        
        send_simple_email(email, subject, body)
        logger.info(f"Email duplicate payment notification sent to {email}")
        
    except Exception as e:
        logger.error(f"Error sending email duplicate payment notification: {e}")

def send_whatsapp_duplicate_notification(phone, bl_number, payment_amount, payment_source):
    """Send WhatsApp notification to user about duplicate payment"""
    if not phone:
        logger.info("No phone available for WhatsApp duplicate payment notification")
        return
        
    try:
        # This would integrate with your WhatsApp API
        # For now, we'll log the message that would be sent
        message = f"⚠️ Duplicate Payment Alert: Your payment of ${payment_amount:.2f} for BL {bl_number} has already been processed. No refund needed."
        
        # TODO: Implement actual WhatsApp sending
        # sendWhatsAppMessage(phone, message)
        
        logger.info(f"WhatsApp duplicate payment notification would be sent to {phone}: {message}")
        
    except Exception as e:
        logger.error(f"Error sending WhatsApp duplicate payment notification: {e}")

def send_staff_refund_alert(bl_number, customer_username, payment_amount, payment_source):
    """Send alert to staff about potential refund request"""
    try:
        # Get staff email addresses
        staff_emails = get_staff_emails()
        
        if not staff_emails:
            logger.warning("No staff emails found for refund alert")
            return
            
        subject = f"🚨 Duplicate Payment Alert - BL {bl_number}"
        
        body = f"""
**Duplicate Payment Detected**

A customer has attempted a duplicate payment that was automatically prevented.

**Details:**
- Bill of Lading: {bl_number}
- Customer: {customer_username}
- Duplicate Amount: ${payment_amount:.2f}
- Payment Source: {payment_source.title()}
- Detection Time: {datetime.now(pytz.timezone('Asia/Hong_Kong')).strftime('%Y-%m-%d %H:%M:%S HKT')}

**Action Required:**
- Monitor for customer support requests
- Be prepared to explain the duplicate payment prevention
- No refund processing needed (payment was not charged)

**Customer Notifications Sent:**
- ✅ FCM Push Notification
- ✅ Email Notification  
- ✅ WhatsApp Message (if configured)

This is an automated alert. The system has already notified the customer.
        """
        
        # Send to all staff members
        for staff_email in staff_emails:
            try:
                send_simple_email(staff_email, subject, body)
                logger.info(f"Staff refund alert sent to {staff_email}")
            except Exception as e:
                logger.error(f"Error sending staff alert to {staff_email}: {e}")
                
    except Exception as e:
        logger.error(f"Error sending staff refund alert: {e}")

def get_staff_emails():
    """Get all staff email addresses"""
    try:
        conn = get_db_conn()
        cur = conn.cursor()
        cur.execute("SELECT customer_email FROM users WHERE role = 'staff' AND approved = true")
        results = cur.fetchall()
        cur.close()
        conn.close()
        
        emails = []
        for row in results:
            if row[0]:
                try:
                    decrypted_email = decrypt_sensitive_data(row[0])
                    emails.append(decrypted_email)
                except Exception as e:
                    logger.error(f"Error decrypting staff email: {e}")
                    
        return emails
    except Exception as e:
        logger.error(f"Error getting staff emails: {e}")
        return [] 

File: whatsapp1/utils/duplicate_payment_notifications.js
/**
 * Duplicate Payment Notification System for WhatsApp App
 * Handles notifications when duplicate payments are detected
 */

const { Pool } = require('pg');
require('dotenv').config();

const pool = new Pool({
  host: process.env.RAILWAY_DB_HOST,
  user: process.env.RAILWAY_DB_USER,
  password: process.env.RAILWAY_DB_PASSWORD,
  database: process.env.RAILWAY_DB_NAME,
  port: process.env.RAILWAY_DB_PORT,
  ssl: false,
});

/**
 * Send comprehensive duplicate payment notifications
 * @param {number} blId - Bill of lading ID
 * @param {string} blNumber - Bill of lading number
 * @param {string} customerUsername - Customer username
 * @param {string} customerEmail - Customer email (encrypted)
 * @param {number} paymentAmount - Amount of duplicate payment
 * @param {string} paymentSource - Source of payment (webhook, email, bank_import, whatsapp)
 * @param {Date} originalPaymentDate - Date of original payment (optional)
 */
async function sendDuplicatePaymentNotifications(blId, blNumber, customerUsername, customerEmail, 
                                               paymentAmount, paymentSource, originalPaymentDate = null) {
  try {
    console.log(`[DUPLICATE PAYMENT] Sending notifications for BL ${blNumber} (ID: ${blId})`);
    
    // Record duplicate payment in unmatched_receipts table for dashboard visibility
    await recordDuplicatePayment(blId, blNumber, customerUsername, paymentAmount, paymentSource, originalPaymentDate);
    
    // Get customer phone number for WhatsApp
    const customerPhone = await getCustomerPhone(customerUsername);
    
    // Get customer FCM tokens
    const fcmTokens = await getCustomerFcmTokens(customerUsername);
    
    // Send notifications
    await sendFcmDuplicateNotification(fcmTokens, blNumber, paymentAmount, paymentSource);
    await sendEmailDuplicateNotification(customerEmail, blNumber, paymentAmount, paymentSource, originalPaymentDate);
    await sendWhatsappDuplicateNotification(customerPhone, blNumber, paymentAmount, paymentSource);
    await sendStaffRefundAlert(blNumber, customerUsername, paymentAmount, paymentSource);
    
    console.log(`[DUPLICATE PAYMENT] Notifications sent for BL ${blNumber} (ID: ${blId})`);
    
  } catch (error) {
    console.error(`[DUPLICATE PAYMENT] Error sending notifications for BL ${blNumber}:`, error);
  }
}

/**
 * Record duplicate payment in unmatched_receipts table for dashboard visibility
 * @param {number} blId - Bill of lading ID
 * @param {string} blNumber - Bill of lading number
 * @param {string} customerUsername - Customer username
 * @param {number} paymentAmount - Payment amount
 * @param {string} paymentSource - Payment source
 * @param {Date} originalPaymentDate - Original payment date (optional)
 */
async function recordDuplicatePayment(blId, blNumber, customerUsername, paymentAmount, paymentSource, originalPaymentDate = null) {
  try {
    // Format original payment date
    let originalDateStr = "";
    if (originalPaymentDate) {
      if (typeof originalPaymentDate === 'string') {
        originalDateStr = ` (Original: ${originalPaymentDate})`;
      } else {
        originalDateStr = ` (Original: ${originalPaymentDate.toISOString().replace('T', ' ').substring(0, 19)})`;
      }
    }
    
    // Create description
    const description = `Duplicate payment for BL ${blNumber} from ${customerUsername}`;
    const reason = `Duplicate Payment: Payment of $${paymentAmount.toFixed(2)} already processed${originalDateStr}`;
    
    // Insert into unmatched_receipts table
    await pool.query(`
      INSERT INTO unmatched_receipts (date, description, amount, reason, raw_text)
      VALUES ($1, $2, $3, $4, $5)
    `, [
      new Date(),
      description,
      paymentAmount,
      reason,
      `Duplicate payment detected for BL ${blNumber} via ${paymentSource}. Customer: ${customerUsername}`
    ]);
    
    console.log(`[DUPLICATE PAYMENT] Recorded duplicate payment in unmatched_receipts for BL ${blNumber}`);
    
  } catch (error) {
    console.error(`[DUPLICATE PAYMENT] Error recording duplicate payment in unmatched_receipts for BL ${blNumber}:`, error);
  }
}

/**
 * Get customer phone number from database
 * @param {string} username - Customer username
 * @returns {Promise<string|null>} - Customer phone number or null
 */
async function getCustomerPhone(username) {
  try {
    const result = await pool.query(
      'SELECT customer_phone FROM users WHERE username = $1',
      [username]
    );
    
    if (result.rows.length > 0 && result.rows[0].customer_phone) {
      // Note: In a real implementation, you would decrypt this
      return result.rows[0].customer_phone;
    }
    return null;
  } catch (error) {
    console.error(`[DUPLICATE PAYMENT] Error getting customer phone for ${username}:`, error);
    return null;
  }
}

/**
 * Get customer FCM tokens from database
 * @param {string} username - Customer username
 * @returns {Promise<string[]>} - Array of FCM tokens
 */
async function getCustomerFcmTokens(username) {
  try {
    const result = await pool.query(
      'SELECT token FROM fcm_tokens WHERE username = $1 AND is_active = true',
      [username]
    );
    
    return result.rows.map(row => row.token);
  } catch (error) {
    console.error(`[DUPLICATE PAYMENT] Error getting FCM tokens for ${username}:`, error);
    return [];
  }
}

/**
 * Send FCM push notification to user about duplicate payment
 * @param {string[]} fcmTokens - Array of FCM tokens
 * @param {string} blNumber - Bill of lading number
 * @param {number} paymentAmount - Payment amount
 * @param {string} paymentSource - Payment source
 */
async function sendFcmDuplicateNotification(fcmTokens, blNumber, paymentAmount, paymentSource) {
  if (!fcmTokens || fcmTokens.length === 0) {
    console.log('[DUPLICATE PAYMENT] No FCM tokens available for notification');
    return;
  }
  
  try {
    const title = "⚠️ Duplicate Payment Detected";
    const body = `Your payment of $${paymentAmount.toFixed(2)} for BL ${blNumber} has already been processed. No action needed.`;
    
    const data = {
      type: 'duplicate_payment',
      bl_number: blNumber,
      payment_amount: paymentAmount.toString(),
      payment_source: paymentSource,
      timestamp: new Date().toISOString()
    };
    
    // Note: In a real implementation, you would call your FCM service here
    console.log(`[DUPLICATE PAYMENT] FCM notification would be sent to ${fcmTokens.length} tokens:`, {
      title,
      body,
      data
    });
    
  } catch (error) {
    console.error('[DUPLICATE PAYMENT] Error sending FCM notification:', error);
  }
}

/**
 * Send email notification to user about duplicate payment
 * @param {string} email - Customer email
 * @param {string} blNumber - Bill of lading number
 * @param {number} paymentAmount - Payment amount
 * @param {string} paymentSource - Payment source
 * @param {Date} originalPaymentDate - Original payment date
 */
async function sendEmailDuplicateNotification(email, blNumber, paymentAmount, paymentSource, originalPaymentDate = null) {
  if (!email) {
    console.log('[DUPLICATE PAYMENT] No email available for notification');
    return;
  }
  
  try {
    const subject = "⚠️ Duplicate Payment Alert - No Action Required";
    
    // Format original payment date
    let originalDateStr = "";
    if (originalPaymentDate) {
      if (typeof originalPaymentDate === 'string') {
        originalDateStr = `Original payment was processed on: ${originalPaymentDate}`;
      } else {
        originalDateStr = `Original payment was processed on: ${originalPaymentDate.toISOString()}`;
      }
    }
    
    const body = `
Dear Customer,

We have detected a duplicate payment attempt for your shipment.

**Payment Details:**
- Bill of Lading: ${blNumber}
- Duplicate Amount: $${paymentAmount.toFixed(2)}
- Payment Source: ${paymentSource.charAt(0).toUpperCase() + paymentSource.slice(1)}
${originalDateStr}

**Important:** Your original payment has already been processed successfully. This duplicate payment will not be charged to your account.

**No action is required from you.** Your shipment processing continues as normal.

If you have any questions or concerns, please contact our support team.

Best regards,
Terry Ray Logistics Team
    `;
    
    // Note: In a real implementation, you would call your email service here
    console.log(`[DUPLICATE PAYMENT] Email notification would be sent to ${email}:`, {
      subject,
      body: body.trim()
    });
    
  } catch (error) {
    console.error('[DUPLICATE PAYMENT] Error sending email notification:', error);
  }
}

/**
 * Send WhatsApp notification to user about duplicate payment
 * @param {string} phone - Customer phone number
 * @param {string} blNumber - Bill of lading number
 * @param {number} paymentAmount - Payment amount
 * @param {string} paymentSource - Payment source
 */
async function sendWhatsappDuplicateNotification(phone, blNumber, paymentAmount, paymentSource) {
  if (!phone) {
    console.log('[DUPLICATE PAYMENT] No phone available for WhatsApp notification');
    return;
  }
  
  try {
    const message = `⚠️ Duplicate Payment Alert: Your payment of $${paymentAmount.toFixed(2)} for BL ${blNumber} has already been processed. No refund needed.`;
    
    // Note: In a real implementation, you would call your WhatsApp API here
    console.log(`[DUPLICATE PAYMENT] WhatsApp notification would be sent to ${phone}:`, message);
    
  } catch (error) {
    console.error('[DUPLICATE PAYMENT] Error sending WhatsApp notification:', error);
  }
}

/**
 * Send alert to staff about potential refund request
 * @param {string} blNumber - Bill of lading number
 * @param {string} customerUsername - Customer username
 * @param {number} paymentAmount - Payment amount
 * @param {string} paymentSource - Payment source
 */
async function sendStaffRefundAlert(blNumber, customerUsername, paymentAmount, paymentSource) {
  try {
    // Get staff email addresses
    const staffEmails = await getStaffEmails();
    
    if (!staffEmails || staffEmails.length === 0) {
      console.log('[DUPLICATE PAYMENT] No staff emails found for refund alert');
      return;
    }
    
    const subject = `🚨 Duplicate Payment Alert - BL ${blNumber}`;
    
    const body = `
**Duplicate Payment Detected**

A customer has attempted a duplicate payment that was automatically prevented.

**Details:**
- Bill of Lading: ${blNumber}
- Customer: ${customerUsername}
- Duplicate Amount: $${paymentAmount.toFixed(2)}
- Payment Source: ${paymentSource.charAt(0).toUpperCase() + paymentSource.slice(1)}
- Detection Time: ${new Date().toISOString()}

**Action Required:**
- Monitor for customer support requests
- Be prepared to explain the duplicate payment prevention
- No refund processing needed (payment was not charged)

**Customer Notifications Sent:**
- ✅ FCM Push Notification
- ✅ Email Notification  
- ✅ WhatsApp Message (if configured)

This is an automated alert. The system has already notified the customer.
    `;
    
    // Note: In a real implementation, you would call your email service here
    console.log(`[DUPLICATE PAYMENT] Staff alert would be sent to ${staffEmails.length} staff members:`, {
      subject,
      body: body.trim(),
      recipients: staffEmails
    });
    
  } catch (error) {
    console.error('[DUPLICATE PAYMENT] Error sending staff refund alert:', error);
  }
}

/**
 * Get all staff email addresses
 * @returns {Promise<string[]>} - Array of staff email addresses
 */
async function getStaffEmails() {
  try {
    const result = await pool.query(
      'SELECT customer_email FROM users WHERE role = $1 AND approved = $2',
      ['staff', true]
    );
    
    const emails = [];
    for (const row of result.rows) {
      if (row.customer_email) {
        // Note: In a real implementation, you would decrypt this
        emails.push(row.customer_email);
      }
    }
    
    return emails;
  } catch (error) {
    console.error('[DUPLICATE PAYMENT] Error getting staff emails:', error);
    return [];
  }
}

module.exports = {
  sendDuplicatePaymentNotifications,
  getCustomerPhone,
  getCustomerFcmTokens,
  sendFcmDuplicateNotification,
  sendEmailDuplicateNotification,
  sendWhatsappDuplicateNotification,
  sendStaffRefundAlert,
  getStaffEmails
}; 